### ISSUE TITLE
After #15921, there is still unconsistent information and an unexpected restart dialog

### ISSUE THREAD

**Issue by CyrilleB79** [nvaccess/nvda#15927]
The solution provided by #15921 is not satisfactory at all. It does not solves all issues with unexpected restart dialog and creates more confusion about add-ons expected to be installed but actually not.

### Steps to reproduce:
- Install 10dita application module and restart NVDA as needed
- Open the appModule sub-sub-folder of this add-on in Windows Explorer and keep this folder open
- Uninstall 10dita and restart NVDA as required
  A1: The add-on's folder is still there and NVDA has not warned the user that the add-on could not be uninstalled.
  E1: If nothing is said, the add-on is expected to be uninstalled successfully, else something should be said to the user.
- Install 10dita and restart
- Open Add-on store:
  A2: The add-on is pending remove
  E2: The add-on is installed
- Close the add-on store
  A3: A dialog indicates that changes were made and that a restart is required.
  E3: No such dialog since no change were made.
- Close Windows Explorer with the appModule folder open
- Press Yes in the restart dialog
  A4: After restart, 10dita is not installed
  E4: After restart, 10dita should be installed since the add-on installation was the last action of the user and the user has not been warned that an action has failed.

There are probably more scenarios causing issues, such as updating an add-on keeping a folder of the old add-on open in Windows Explorer.

### Actual behavior:
See A1, A2, A3 and A4 in STR
### Expected behavior:
See E1, E2, E3 and E4 in STR.

More generally, I think that the add-on store should not consider add-ons uninstalled if they have not been installed successfully. The real state (pending remove here) of the add-on should be displayed.

On the other side with #15719, if installation fails, NVDA should not start normally and remove silently the add-on. Unless I am mistaken, the root cause of #1571 has not been addressed in #15921 (nor found), so failing silently will not help to find this root cause.

Maybe the solution of @lukaszgo1 in https://github.com/nvaccess/nvda/pull/15921#issuecomment-1855626467 can be of interest.
### NVDA logs, crash dumps and other attachments:
None
#### NVDA installed/portable/running from source:
From source
#### NVDA version:
Last beta branch, commit 81e17be07bbebe02f682fbc82c23c698b865c50f.
#### Windows version:
Windows 10 22H2 (AMD64) build 19045.3693
#### Name and version of other software in use when reproducing the issue:
N/A
#### Other information about your system:
N/A
#### Does the issue still occur after restarting your computer?
Not tested.
#### Have you tried any other versions of NVDA? If so, please report their behaviors.
N/A, (#15921 just merged.

#### If NVDA add-ons are disabled, is your problem still occurring?
N/A
#### Does the issue still occur after you run the COM Registration Fixing Tool in NVDA's tools menu?
Not tested

**Comment by lukaszgo1** [comment]
Would you consider pre #15921 behavior to be acceptable for 2024.1? In other words would it be okay if NVDA would not remove failed installs, and continue attempting them on the next restart? this is not to say that #12823 should not be looked into, but lack of warnings displayed to the user when add-on installs fails is not a new problem. I certainly agree that #15719 should not have been closed, as we do not know what caused it ,nor if it is actually fixed. My preference would be to submit a PR which reverts #15921 and removes add-ons which are no longer on the file system from the pending installs set. The remaining issues with add-ons failing to be installed / removed should be solved separately. I'm also curious how often this occurs in real life situations, i.e. not with a toy example where someone keeps the add-on folder open in Windows Explorer just to demonstrate the problem.

**Comment by CyrilleB79** [comment]
> Would you consider pre #15921 behavior to be acceptable for 2024.1? In other words would it be okay if NVDA would not remove failed installs, and continue attempting them on the next restart? 

Yes, it's acceptable for my use case (#12629).

> this is not to say that #12823 should not be looked into, but lack of warnings displayed to the user when add-on installs fails is not a new problem. 

I agree, this can be deferred to NVDA 2024.2, or whenever I or anybody else bother to implement it.

> I certainly agree that #15719 should not have been closed, as we do not know what caused it ,nor if it is actually fixed. My preference would be to submit a PR which reverts #15921 and removes add-ons which are no longer on the file system from the pending installs set.

Yes, with the appropriate logging.


> The remaining issues with add-ons failing to be installed / removed should be solved separately. I'm also curious how often this occurs in real life situations, i.e. not with a toy example where someone keeps the add-on folder open in Windows Explorer just to demonstrate the problem.

#12629 was not a toy example. I was facing it regularly when working on add-on development. It's common to work on an installed add-on, and when the code is ready, copy it to Git folder, generate the add-on with `scons` and try to install it to check the final version.

**Comment by paulber19** [comment]
For the last point, I also face this problem regularly.
To maintain the compatibility of my add-ons with several versions of 
NVDA, I test on a portable version, make modifications, create the 
add-on then change the portable version and install the new version of 
the add-on on this version.
No more files from the previous version of the add-on are used.
Very often, the installation goes well, but upon reboot, an error is in 
the log. the new version of the add-on is not installed because the 
folder of the previous version could not be deleted.

**Comment by lukaszgo1** [comment]
Re: the last point. I'm not sure if I fully understand the described scenario. When working on add-ons I try to follow one of the two workflows:
1. Write the new code on the copy of the add-on in the GIT repository, when ready compile it with `scons`, install into the running NVDA, restart and repeat as needed, until the end result is satisfactory or
2. Work on the copy of the add-on in NVDA's config until the result is okay, when I'm satisfied copy it into the repository, compile with `scons` and install into the running NVDA

At which point you have to have add-on folder open in Windows Explorer when installing? Or is it just that it is easy to forget that it should be closed?

**Comment by lukaszgo1** [comment]
@CyrilleB79  I started working on this, but looking at your STR some things are not very clear for me. Would you be able to help me clarify the following points?
> 
> * Install 10dita and restart

Should it be even possible to install an add-on which is still pending remove from the available tab in your opinion?

> 
> * Open Add-on store:
>       A2: The add-on is pending remove
>       E2: The add-on is installed

Isn't this a mistake, if the old version of the add-on cannot be uninstalled, which indirectly means that the new installation cannot succeed either, the expected status should be 'pending remove' if the version which was attempted to be uninstalled has preference, or 'pending install' if the status of the newly installed build should be preferred.

**Comment by CyrilleB79** [comment]
> Re: the last point. I'm not sure if I fully understand the described scenario. When working on add-ons I try to follow one of the two workflows:
> 
>     1. Write the new code on the copy of the add-on in the GIT repository, when ready compile it with `scons`, install into the running NVDA, restart and repeat as needed, until the end result is satisfactory or
> 
>     2. Work on the copy of the add-on in NVDA's config until the result is okay, when I'm satisfied copy it into the repository, compile with `scons` and install into the running NVDA
> 
> 
> At which point you have to have add-on folder open in Windows Explorer when installing? Or is it just that it is easy to forget that it should be closed?

I usually follow scenario 2. It is not mandatory to have Windows Explorer open; it's just that I use it to open files from NVDA config folder and then I sometimes forget to close it before reinstalling the add-on.


> Should it be even possible to install an add-on which is still pending remove from the available tab in your opinion?

I do not expect nor request this at all, since it was not the case until now, and I guess that I creates more difficulty. Anyway, if you have a solution to this, I am not opposed to it.

> > * Open Add-on store:
> >   A2: The add-on is pending remove
> >   E2: The add-on is installed
> 
> Isn't this a mistake, if the old version of the add-on cannot be uninstalled, which indirectly means that the new installation cannot succeed either, the expected status should be 'pending remove' if the version which was attempted to be uninstalled has preference, or 'pending install' if the status of the newly installed build should be preferred.

But I have not been warned that the add-on could not be uninstalled. Thus, at this point, I believe that I have uninstalled the add-on successfully and then also reinstalled it successfully.

**Comment by CyrilleB79** [comment]
Thanks @lukaszgo1, your branch seems interesting.

I have done some tests but not exhaustive tests. My first feedback:
* When the add-on cannot be uninstalled due to locked resources, e.g. folder open in Windows Explorer, the error message is appreciated. We can still see in the Add-on Store that the message is pending remove, which is the real status of this add-on.
* However if you go to available tab at this moment, you can pick up an update if available, what should not be allowed IMO. I do not know if this issue was already in NVDA beta branch or if it's new in your branch.

**Comment by lukaszgo1** [comment]
> * However if you go to available tab at this moment, you can pick up an update if available, what should not be allowed IMO. I do not know if this issue was already in NVDA beta branch or if it's new in your branch.



This issue is already present in the beta branch and therefore I consider it to be out of scope. I have created #15937 - if you have time could you please test / review / mention what additional tests should be performed in your opinion?
